### PR TITLE
[Feature] Bulk delete users

### PR DESCRIPTION
**Pull request by noliveleger**
## Description 

Allow the service account to delete users entirely

## Related issues

kobotoolbox/kpi#4373

### CODE REVIEW

File: onadata/apps/logger/models/xform.py [hunk 1]
```diff
@@ -89,6 +99,7 @@ class XForm(BaseModel):
 
     has_kpi_hooks = LazyDefaultBooleanField(default=False)
     kpi_asset_uid = models.CharField(max_length=32, null=True)
+    pending_delete = LazyDefaultBooleanField(default=False)
```
Comment (jnm) [inline onadata/apps/logger/models/xform.py:102]: Apparently we no longer need this lazy-field stuff starting with [PostgreSQL 11](https://www.postgresql.org/docs/11/sql-altertable.html) :tada:

File: onadata/apps/logger/models/xform.py [hunk 2]
```diff
@@ -47,6 +47,16 @@ def upload_to(instance, filename):
...
 
 
+class XFormWithoutPendingDeletedManager(models.Manager):
+    def get_queryset(self):
+        return super().get_queryset().exclude(pending_delete=True)
+
+
+class XFormAllManager(models.Manager):
+    def get_queryset(self):
+        return super(models.Manager, self).get_queryset()
```
Comment (jnm) [inline onadata/apps/logger/models/xform.py:57]: `pass` :blush:

**Review by jnm — COMMENTED** [review]
Initial read-through while still in progress. Looking good :+1:

File: onadata/apps/main/migrations/0011_drop_old_kpi_tables.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    'koboform_surveydraft',
+    'koboform_surveypreview',
+    'registration_supervisedregistrationprofile',
+    'registration_registrationprofile',
+    'south_migrationhistory',
+    'stats_statscount',
+]
+
+
+def get_operations():
```
[the remark this answers is not in the document]
Comment (noliveleger) [reply onadata/apps/main/migrations/0011_drop_old_kpi_tables.py:54]: As per our discussion, we did improve the migration by dropping the constraints first and drop tables.

File: onadata/apps/api/tests/viewsets/test_xform_viewset.py [hunk 4]
```diff
@@ -64,6 +64,25 @@ def test_form_list_other_user_access(self):
         # should be empty
         self.assertEqual(response.data, [])
 
+    def test_form_list_with_pending_delete_xform(self):
+        """
+        Test that bob (or anyone else) has access to bob's pending delete XForm
```
Comment (jnm) [inline onadata/apps/api/tests/viewsets/test_xform_viewset.py:69]: "does not have access"?

File: onadata/apps/api/viewsets/xform_viewset.py [hunk 5]
```diff
@@ -705,9 +714,23 @@ def csv_import(self, request, *args, **kwargs):
...
+        xform_id_string = instance.id_string
+
+        instance.delete()
+        # Clean up storage
+        default_storage.delete(str(instance.xls))
+        rmdir(f'{username}/form-media/{xform_uuid}')
+        rmdir(f'{username}/exports/{xform_id_string}')
+        # Attachments should have been already deleted. Let's remove the parent
+        # folder entirely anyway in case orphans are still present.
+        rmdir(f'{username}/attachments/{xform_uuid}')
```
Comment (jnm) [inline onadata/apps/api/viewsets/xform_viewset.py:736]: maybe excessively paranoid, but can we add a check to make sure the UUID and `id_string` are not empty before calling `rmdir()`?

**Review by jnm — CHANGES_REQUESTED** [review]

File: onadata/apps/api/tests/viewsets/test_xform_viewset.py (hunk 4 above)
Comment (noliveleger) [reply onadata/apps/api/tests/viewsets/test_xform_viewset.py:69]: Yep, I wanted to say "no access" .

**Pull request merged**